packages/git/src/index.ts

```
@@ -43,13 +45,126 @@ export async function getDivergedCommit(cwd: string, ref: string) {
   return cmd.stdout.toString().trim();
 }
 
-async function getCommitThatAddsFile(gitPath: string, cwd: string) {
-  const gitCmd = await spawn(
-    "git",
-    ["log", "--diff-filter=A", "--max-count=1", "--pretty=format:%h", gitPath],
-    { cwd }
-  );
-  return gitCmd.stdout.toString();
+const getCommitThatAddsFile = deprecate(
+  async (gitPath: string, cwd: string) => {
+    return (await getCommitsThatAddFiles([gitPath], cwd))[0];
+  },
+  "Use the bulk getCommitsThatAddFiles function instead"
+);
+
+/**
+ * Get the short SHAs for the commits that added files, including automatically
+ * extending a shallow clone if necessary to determine any commits.
+ * @param gitPaths - Paths to fetch
+ * @param cwd - Location of the repository
+ */
+async function getCommitsThatAddFiles(
+  gitPaths: string[],
+  cwd: string
+): Promise<(string | undefined)[]> {
+  // Maps gitPath to short commit SHA
+  const map = new Map<string, string>();
+
+  // Paths we haven't completed processing on yet
+  let remaining = gitPaths;
+
+  do {
+    // Fetch commit information for all paths we don't have yet
+    const commitInfos = await Promise.all(remaining.map(findCommitAndParent));
+
+    // To collect commits without parents (usually because they're absent from
+    // a shallow clone).
+    let commitsWithMissingParents = [];
+
+    for (const info of commitInfos) {
+      if (info.commitSha) {
+        if (info.parentSha) {
+          // We have found the parent of the commit that added the file.
+          // Therefore we know that the commit is legitimate and isn't simply the boundary of a shallow clone.
+          map.set(info.path, info.commitSha);
+        } else {
+          commitsWithMissingParents.push(info);
+        }
+      } else {
+        // No commit for this file, which indicates it doesn't exist.
```

could you elaborate on what does it mean? the file? the commit? seems like a file - shouldn't we throw in such a case instead of silently ignoring this?

---

This is the case where we've asked for the commit that introduced a file but the file doesn't exist.  It doesn't silently fail: it returns an `undefined` in that return slot.  So if you ask for the commits for 3 files, and only 2 exist, you'll still get the commits for the 2 that did exist, and `undefined` for the one that didn't exist. (There's a unit test for this.)

I could have made the bulk operation throw if any file didn't exist, but I think the previous implementation simply returned `undefined` for a non-existent file, so I kept that behaviour.  Happy to change it to a throw if any requested files don't exist; I slightly prefer that approach myself.

---

> I could have made the bulk operation throw if any file didn't exist, but I think the previous implementation simply returned undefined for a non-existent file, so I kept that behaviour. 

Right, good point about this being the current behavior. Let's keep it that way for now - but I think we might change this to throwing in the future.